src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py

```diff
@@ -220,6 +225,124 @@ def __init__(
+
+        # if there is no duplicate node, no optimization will be performed
+        duplicate_plan_set = find_duplicate_nodes(self)
+        if not duplicate_plan_set:
+            return self
+
+        def find_duplicate_subtrees(root: SnowflakePlan) -> Set[SnowflakePlan]:
+            """
+            Returns a set containing all roots of duplicate subtrees in query plan tree.
+            The root of a duplicate subtree is defined as a duplicate node whose parent is not
+            duplicate.
+            This function is used to only include nodes that should be converted to a CTE.
+            """
+            result = set()
+
+            def traverse(node: SnowflakePlan, parent: Optional[SnowflakePlan]) -> None:
+                if (
+                    parent is not None
+                    and parent not in duplicate_plan_set
+                    and node in duplicate_plan_set
+                ):
+                    result.add(node)
+
+                if node.source_plan and node.source_plan.children:
+                    for child in node.source_plan.children:
+                        traverse(child, node)
+
+            traverse(root, None)
+            return result
+
+        filtered_duplicate_plan_set = find_duplicate_subtrees(self)
+
+        plan_to_query_map = {}
+        filtered_duplicate_plan_to_cte_map = {}
+        filtered_duplicate_plan_to_table_name_map = {}
+
+        def build_plan_to_query_map_in_poster_order(node: SnowflakePlan) -> None:
+            """
+            Builds a mapping from query plans to queries that are optimized with CTEs,
+            in post-traversal order. We can get the final query from the mapping value of the root node.
+            The reason of using poster-traversal order is that chained CTEs have to be built
+            from bottom (innermost subquery) to top (outermost query).
+            """
+            if not node.source_plan or node in plan_to_query_map:
+                return
+
+            for child in node.source_plan.children:
+                build_plan_to_query_map_in_poster_order(child)
+
+            if not node.placeholder_query:
+                plan_to_query_map[node] = node.queries[-1].sql
+            else:
+                plan_to_query_map[node] = node.placeholder_query
+                for child in node.source_plan.children:
+                    # replace the placeholder (id) with child query
+                    plan_to_query_map[node] = plan_to_query_map[node].replace(
+                        child._id, plan_to_query_map[child]
+                    )
+
+            # duplicate subtrees will be converted CTEs
+            if node in filtered_duplicate_plan_set:
+                # when a subquery is converted a CTE to with clause,
+                # it will be replaced by `SELECT * from TEMP_TABLE` in the original query
+                table_name = random_name_for_temp_object(TempObjectType.TABLE)
```

Comment:
I'm wondering, whether we should create another object type for CTE, which can be less confusing to customers (because it's not a real table) and easier for us to debug.

Reply:
I would not be opposed to it

Reply:
Yeah I agree.

Reply:
done